src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepositoryTest.kt

```diff
@@ -1,59 +1,312 @@
+    }
+
+    val serviceProvider = serviceProviderFactory.create(random(13), random(14))
+    val contract = when {
+      asPrime -> dynamicFrameworkContractFactory.create(primeProvider = serviceProvider, contractReference = random(10))
+      else -> dynamicFrameworkContractFactory.create(subcontractorProviders = setOf(serviceProvider), contractReference = random(10))
+    }
+    val intervention = interventionFactory.create(contract = contract)
+    val referral = referralFactory.createSent(
+      intervention = intervention,
+      assignments = assignedUsers.map { ReferralAssignment(OffsetDateTime.now(), assignedBy = it, assignedTo = it) }
+    )
+    val serviceUser = serviceUserFactory.create(random(15), random(16), referral)
+    return Multiple(referral.currentAssignee, serviceProvider, contract, intervention, referral, serviceUser)
+  }
+
+  private fun createEndedReferral(
+    asPrime: Boolean,
+    endRequestedAt: OffsetDateTime? = null,
+    concludedAt: OffsetDateTime? = null,
+    eosR: EndOfServiceReport? = null,
+  ): Multiple {
+
+    val user = authUserFactory.create(random(6), random(5), random(12))
+    val serviceProvider = serviceProviderFactory.create(random(13), random(14))
+    val contract = when {
+      asPrime -> dynamicFrameworkContractFactory.create(primeProvider = serviceProvider, contractReference = random(10))
+      else -> dynamicFrameworkContractFactory.create(subcontractorProviders = setOf(serviceProvider), contractReference = random(10))
+    }
+    val intervention = interventionFactory.create(contract = contract)
+    val referral = referralFactory.createEnded(
+      intervention = intervention,
+      assignments = listOf(ReferralAssignment(OffsetDateTime.now(), assignedBy = user, assignedTo = user)),
+      endRequestedAt = endRequestedAt,
+      concludedAt = concludedAt,
+      endOfServiceReport = eosR
+    )
+    val serviceUser = serviceUserFactory.create(random(15), random(16), referral)
+    return Multiple(referral.currentAssignee, serviceProvider, contract, intervention, referral, serviceUser)
+  }
+
+  private fun createDraftReferral(
+    asPrime: Boolean
+  ): Multiple {
+
+    val serviceProvider = serviceProviderFactory.create(random(13), random(14))
+    val contract = when {
+      asPrime -> dynamicFrameworkContractFactory.create(primeProvider = serviceProvider, contractReference = random(10))
+      else -> dynamicFrameworkContractFactory.create(subcontractorProviders = setOf(serviceProvider), contractReference = random(10))
+    }
+    val intervention = interventionFactory.create(contract = contract)
+    val referral = referralFactory.createDraft(intervention = intervention)
+    val serviceUser = serviceUserFactory.create(random(15), random(16), referral)
+    return Multiple(referral.currentAssignee, serviceProvider, contract, intervention, referral, serviceUser)
+  }
+
+  private fun random(length: Int): String {
```

Comment:
I personally don't like the use of random in tests; I think testing should be done with defined constraints, randomness introduces uncertainty.

Reply:
The actual value of a service provider id adds no value to the test so rather than make them up I used random strings